Do not execute required mock

royriojas opened the pull request:
Hi guys this is the improvement proposed as part of https://github.com/thlorenz/proxyquireify/issues/38

Couple of notes.

- the magic just wraps the call to require with an anonymous function. Since it is not executed it prevents double calls to require that were happening before.
- the tests were consistently failing, some sort of race condition maybe. I just solved it splitting the problematic test into its own module and launching it after the other tests. Would be really nice to figure it out what went wrong, but at least this helps pass the tests for now

TheSavior commented:
I'd be surprised if this got merged. It clearly isn't the right solution to have to run it separately. I'm not familiar enough with the code either to know what is actually happening.

royriojas commented:
The fix for the tests is not what I'm interested about, but for the one avoiding the `require` calls for mocks to be executed. I just wanted to be sure the code I introduced was not the guilty of the failures, as it clearly wasn't. The failure happens on browserify > 10.2.4. So it is probably something on their end and not it the tests

royriojas commented:
I have found the way to fix the tests, it seems that there is some sort of race condition. The approach I have here is first compile the code, then run the tests, with that, it seems that the race condition went away and no need to have one of the tests to run alone

thlorenz commented on package.json:
```diff
@@ -23,6 +23,7 @@
   },
   "devDependencies": {
     "browserify": "^10.0.0",
+    "es6-promise": "^2.3.0",
```
why do we need this?

royriojas replied:
The tests were broken (even without any changes) for browserify > 10.2.4 the change that I made was to make sure the tests were run sequentially without enter into race conditions which happened to fix the issues. I haven't change any other test, just make sure they were properly run in sequence

thlorenz commented:
Please don't change any existing tests or how they are run.
Otherwise we cannot be sure that no existing functionality is affected by your change.
Thanks.

royriojas commented:
@thlorenz this change required to change 1 of the tests because it no longer make sense. Since before you just put a require call the code of the module was executed immediately. this reported that the module was required 3 times, when in effect it was only required to be included twice.

The tests were broken (even without any changes) for browserify > 10.2.4 the change that I made was to make sure the tests were run sequentially without enter into race conditions which happened to fix the issues. I haven't change any other test, just make sure they were properly run in sequence

royriojas commented on test/clientside/independent-overrides.js:
```diff
@@ -10,8 +10,8 @@ var foober =  proxyquire('../fixtures/foo', { './bar': barber });
 var foo    =  proxyquire('../fixtures/foo', { './bar': { } });
 
 test('\noverriding bar.bar for foober but not for foo', function (t) {
-  t.equal(window.foostats.fooRequires(), 3, 'foo is required three times since one for each test and one for require detective')
```
This is the only test I change, because as I mentioned you were incorrectly expecting the module to be required 3 times, 1 for the magic require, and 2 for the require calls.

thlorenz replied:
Ok, but this test had to change due to this behavior change right?
So it wasn't wrong (not wrong now either), but having to change it reflects that we broke backwards compat.
Some people's tests may depend on the previous behavior.
So we'll need to bump major version for this (or at a minimum minor).

royriojas commented on test/clientside/run.js:
```diff
@@ -5,27 +5,51 @@ var browserify =  require('browserify');
 var proxyquire =  require('../..');
 var vm         =  require('vm');
 var test       =  require('tape');
+var ES6Promise =  require('es6-promise').Promise;
```
This was just used to make sure that all the tests were compiled before attempt to run tape, to avoid the race conditions

bendrucker commented:
No need for promises to coordinate callback. run-parallel would do just fine on first glance.

royriojas commented:
@sure, I'm not familiar with `run-parallel` if you point me to an example I can make the change if required.

bendrucker commented:
https://github.com/search?q=user%3Afeross+run-

thlorenz commented:
Let's do this w/out adding any dependencies like this unless we have a good reason.
A simple callback should work here.

Additionally don't change anything about the way the tests run please.
I'm not sure why that would be required to add the feature, so you'd have to have a good reason and explain why you need this.

thlorenz commented:
An example to wait for multiple parallel things to happen is [here](http://thlorenz.com/blog/handling-multiple-failing-async-requests/).

royriojas commented:
@thlorenz sorry I updated the pull request before reading you don't want new dependencies. I used what @bendrucker suggested. `run-parallel`

So let's me explain why the test runner was modified.

1. If you pull your repo from `master` and run `npm test` you will see the tests are failing for browserify > 10.2.4
2. the issue seems to be that there is a race condition when entering the part that executes the code. 
3. I realize that the first part of your code was actually creating the bundle, so in order to make things simpler, I moved the bundling part from the actual execution of the tests.

let me be clear that I haven't modified any tests, but the one that was incorrect: expecting the require to happen 3 times was not correct if the code only 2 calls to proxyquire were done. The extra one was for the magic require inserted.

Let me know if you want me to modify this to not use any dep, I will kindly ask you to consider `run-parallel` or even promises since they are dev dependencies not real dependencies.

Thanks for your time guys

thlorenz commented on lib/transform.js:
```diff
@@ -8,8 +8,9 @@ var through          =  require('through')
-    requireDeps(deps);
+
+  return '/* proxyquireify injected requires to make browserify include dependencies in the bundle */ /* istanbul ignore next */; (function() { ' +
+    requireDeps(deps) + '});\n\n';
```
Please name this function like `makeBrowserifyIncludeModule` or similar, so that people looking at the bundle have an idea what's going on, especially if they see these functions in their heap dumps.

thlorenz commented on lib/transform.js:
```diff
@@ -18,7 +19,7 @@ module.exports = function (file) {
 
   return through(write, end);
-    
+
```
Please remove whitespace changes

thlorenz commented on test/clientside/run.js:
```diff
@@ -5,27 +5,49 @@ var browserify =  require('browserify');
+  .bundle()
+  .on('error', function error(err) { cb(err); })
+  .on('data', function (data) { src += data })
+  .on('end', function () {
```
Please undo these indent changes.

thlorenz commented on test/clientside/run.js:
```diff
@@ -5,27 +5,49 @@ var browserify =  require('browserify');
+  'independent-overrides'
+  , 'manipulating-overrides'
+  , 'noCallThru'
+  , 'argument-validation'
```
Please fix indent.

thlorenz commented:
Overall LGTM. Please fix the places I pointed out.

> the issue seems to be that there is a race condition when entering the part that executes the code

We should check that that's not exposing a problem we need to fix in the code instead of just modifying the tests to cover it up.

royriojas commented:
Hi @thlorenz,

I guess I have cover all of your observations. I do agree, this might be at least a minor or even major change, since it changes one of the unit tests. 

I have removed `run-parallel`.

Thanks for your time

thlorenz commented on lib/transform.js:
```diff
@@ -8,8 +8,8 @@ var through          =  require('through')
-  return '/* proxyquireify injected requires to make browserify include dependencies in the bundle */;' +
-    requireDeps(deps);
+  return '/* proxyquireify injected requires to make browserify include dependencies in the bundle */ /* istanbul ignore next */; (function __makeBrowserifyIncludeModule__() { ' +
+    requireDeps(deps) + '});\n\n';
```
Please remove `\n\n` since we don't want to change the user's code too much. 
Injecting a line right in front of the other was done to preserve line numbers between code that runs with and without proxyquireify.

thlorenz commented on test/clientside/run.js:
```diff
@@ -15,17 +14,54 @@ function run(name) {
+};
+
+var clientTests = [
+  'independent-overrides'
```
Indent here still broken, please line up all strings.

royriojas replied:
something like this?

```javascript
var clientTests = [
    'independent-overrides'
  , 'manipulating-overrides'
  , 'noCallThru'
  , 'argument-validation'
];

```

thlorenz replied:
Yes.

thlorenz commented:
LGTM once the last few nits are fixed.
Gonna wait for @bendrucker to review before merging.

bendrucker commented:
Made some small stylistic comments, LGTM once they're resolved

thlorenz commented on test/clientside/run.js:
```diff
@@ -15,17 +14,54 @@ function run(name) {
     .require(require.resolve('./' + name), { entry: true })
     .bundle()
-    .on('error',  function error(err) { console.error(err); process.exit(1); })
+    .on('error', function error(err) { cb(err); })
```
Here you could just do ` .on('error', cb)` instead.

royriojas replied:
sure

royriojas replied:
done

royriojas commented:
All but the comment for other coverage tools are resolved. Each coverage tool has different ways to ignore block of codes.

I imagine that maybe a solution would be to actually register a module and the possible aliases during bundling, but that would require going back to the drawing board. I guess for now this is good enough

bendrucker commented:
I don't really use coverage tools these days. Can you give some examples?

royriojas commented:
Hi @bendrucker, 

the comment instruction I put there was for `babel-istanbul`, I used with [browserify-babel-istanbul](https://github.com/malandrew/browserify-babel-istanbul) they use /* istanbul ignore next */

I couldn't find info about any other coverage tool, `blanketjs` is the other one I know, but that doesn't have a way to ignore the generated code.

bendrucker commented:
Ok, I think it's fine to leave it for now

thlorenz commented:
Noted some more minor nits in last last commit, please fix and we can merge if @bendrucker is ok with everything.

BTW thanks for your patience with nit fixing, but we need to uphold current style in module.

royriojas commented:
@thlorenz I made the requested changes. BTW Thanks for writing the module in the first place.

bendrucker commented:
Looks okay to me, pending a squash

royriojas commented:
done!

thlorenz commented:
So is this good to merge @bendrucker? If yes I'll try to get to it tomorrow.

bendrucker commented:
Yes!

thlorenz commented:
Ok, merged 0d5fe82f00b2a9e12c138f59544569cae1fb220d and published as v3.
This is a major upgrade as discusses due to the breaking change in behavior.

bendrucker commented:
@royriojas Can you do me a favor and send a PR to proxyquire-universal updating the peerDep?

royriojas commented:
sure @bendrucker, will do.

royriojas commented:
@bendrucker It seems that `proxyquire-universal` was pin to use 1.x.x versions of `proxyquireify`. Tests fails if I even try to use version 2.0.0 of `proxyquireify`. That will require change on the tests on `proxyquire-universal`. Are you open to do that? Wouldn't it be better to create a new issue for the update?

bendrucker commented:
Yeah sure, really appreciate it

thlorenz commented:
@royriojas I'd just create a PR over there with what you have so far (think of a PR as an issue with code).
Then you guys can figure out how to solve the problem.

royriojas commented:
@thlorenz ok will open then the PR. Test won't pass though
